src/block/request.rs

```diff
@@ -0,0 +1,244 @@
+        }
+    }
+}
+
+/// Dedicated [`Result`](https://doc.rust-lang.org/std/result/) type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// Type of request from driver to device.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RequestType {
+    /// Read request.
+    In,
+    /// Write request.
+    Out,
+    /// Flush request.
+    Flush,
+    /// Unknown request.
+    Unsupported(u32),
+}
+
+impl From<u32> for RequestType {
+    fn from(value: u32) -> Self {
+        match value {
+            VIRTIO_BLK_T_IN => RequestType::In,
+            VIRTIO_BLK_T_OUT => RequestType::Out,
+            VIRTIO_BLK_T_FLUSH => RequestType::Flush,
+            t => RequestType::Unsupported(t),
+        }
+    }
+}
+
+/// Block request header.
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C)]
+struct RequestHeader {
+    request_type: u32,
+    _reserved: u32,
+    sector: u64,
+}
+
+/// Stores the necessary information for further execution of a block request.
+#[derive(Debug, PartialEq)]
+pub struct Request {
+    /// The type of the request.
+    request_type: RequestType,
+    /// Stores the (address, data length) pairs where the data descriptors
+    /// point to.
+    data: Vec<(GuestAddress, u32)>,
+    /// The offset (multiplied by 512) where the read or write is to occur.
+    sector: u64,
+    /// The address where the device should write the request status.
+    status_addr: GuestAddress,
+}
+
+// Safe because RequestHeader contains only plain data.
+unsafe impl ByteValued for RequestHeader {}
+
+impl Request {
+    /// Returns the request type.
+    pub fn request_type(&self) -> RequestType {
+        self.request_type
+    }
+
+    /// Returns the (address, len) pairs where the request data is in the guest memory.
+    pub fn data(&self) -> &[(GuestAddress, u32)] {
+        &self.data
+    }
+
+    /// Returns the sector.
+    pub fn sector(&self) -> u64 {
+        self.sector
+    }
+
+    /// Returns the status address.
+    pub fn status_addr(&self) -> GuestAddress {
+        self.status_addr
+    }
+
+    // Checks that a descriptor meets the minimal requirements for a valid status descriptor.
+    fn check_status_desc<M: GuestAddressSpace>(mem: &<M>::M, desc: Descriptor) -> Result<()> {
+        // The status MUST always be writable.
+        if !desc.is_write_only() {
+            return Err(Error::UnexpectedReadOnlyDescriptor);
+        }
+        // It makes sense to check the data length from descriptor, since we're going to
+        // keep from now on only the address of the status buffer in the `Request` object
+        // so we won't be able to validate this at some other point.
+        if desc.len() < 1 {
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        // Check that the address of the status descriptor is valid in guest memory.
+        // We will write an u8 status here after executing the request.
+        let _ = mem
```

Comment:
As mentioned in another comment, I would leave these checks so that we don't have partial execution of requests before the error is thrown (for read, write) or entire request execution for the case when the status address is invalid.